src/index.ts

```diff
@@ -40,78 +29,72 @@ async function bundleMDX(
-    name: 'in-memory-module',
-    resolveId(importee: string, importer?: string) {
-      if (!importer || importee === entryPath) return importee
-      if (!importee[0].startsWith('.')) return null
-
-      const resolved = path.resolve(path.dirname(importer), importee)
-
-      if (resolved in absoluteFiles) return resolved
-      for (const ext of ['.js', '.tsx', '.jsx', '.ts']) {
-        const resolvedWithExt = `${resolved}${ext}`
-        if (resolvedWithExt in absoluteFiles) return resolvedWithExt
-      }
-      throw new Error(
-        `Could not resolve '${importee}' from ${
-          importer === entryPath
-            ? 'the entry MDX file'
-            : `'${importer.replace(`${dir}/`, '')}'`
-        }`,
-      )
-    },
-    load(id: string) {
-      if (absoluteFiles[id]) return absoluteFiles[id]
-      return null
+
+  const inMemoryPlugin: Plugin = {
+    name: 'inMemory',
+    setup(build) {
+      build.onResolve({filter: /.*/}, args => {
+        if (args.path === entryPath) return {path: args.path}
+
+        const modulePath = path.resolve(path.dirname(args.importer), args.path)
+
+        if (modulePath in absoluteFiles) return {path: modulePath}
+
+        for (const ext of ['.js', '.ts', '.jsx', '.tsx']) {
+          const fullModulePath = `${modulePath}${ext}`
+          if (fullModulePath in absoluteFiles) return {path: fullModulePath}
+        }
+
+        console.dir(args.importer)
+
+        throw new Error(
+          `Could not resolve ${args.path} in ${
+            args.importer === entryPath
+              ? 'the entry MDX file.'
+              : args.importer.replace(dir, '.')
+          }`,
+        )
+      })
+
+      build.onLoad({filter: /__mdx_bundler_fake_dir__/}, args => {
+        return {
+          contents: absoluteFiles[args.path],
+          loader: 'jsx',
+        }
+      })
     },
   }
 
-  const inputOptions = getInputOptions({
-    external: ['react', 'react-dom'],
-    input: entryPath,
+  const buildOptions = esbuild({
+    entryPoints: [entryPath],
+    write: false,
     plugins: [
-      inMemoryModulePlugin,
-      nodeResolve(),
-      commonjs({include: 'node_modules/**', sourceMap: false}),
-      json(),
-      rollupBabel({
-        babelHelpers: 'inline',
-        configFile: false,
-        exclude: /node_modules/,
-        extensions: ['.js', '.ts', '.tsx', '.md', '.mdx', '.jsx', '.json'],
-        presets: [
-          [babelPresetReact, {pragma: 'mdx'}],
-          babelPresetEnv,
-          [babelPresetTypeScript, {allExtensions: true, isTSX: true}],
-        ],
-        sourceMaps: false,
+      nodeResolve({
+        extensions: ['.js', '.ts', '.jsx', '.tsx'],
       }),
-      terser(),
+      inMemoryPlugin,
     ],
-  })
-
-  const outputOptions = getOutputOptions({
-    name: 'Component',
+    bundle: true,
+    external: ['react', 'react-dom'],
     format: 'iife',
-    globals: {
-      react: 'React',
-      'react-dom': 'ReactDOM',
-    },
```

Comment:
This is a feature we need to make sure we have. How does esbuild know that references to the `'react'` module should be replaced with a global reference of `React`?

Reply:
It uses the classic runtime: https://esbuild.github.io/content-types/#jsx.

If xdm is added, it’ll work on the MDX files and compile JSX away to use the new automatic runtime (configurable), *before* esbuild kicks in.

But xdm won‘t take care of `.js` files, so esbuild would still compile JSX -> JS there. Not sure if esbuild is interested in adding the new runtime

Reply:
This is sorted in the latest commit.